### PR TITLE
Feature: support for creation, deletion, modification of network segm…

### PR DESCRIPTION
**Pull request by fr0gger03**
…ents and T1 gateways from 1.18 onwards.

Includes new commands for:
configure-t1
remove t1
new-segment
configure-segment
remove-segment

To support bot 'fixed' segments (pre-1.18) and 'moveable' T1 and segments (1.18 and later).

Signed-off-by: Tom Twyman <ttwyman@vmware.com>

### CODE REVIEW

File: pyVMC.py [hunk 1]
```diff
@@ -2189,18 +2190,133 @@ def getSDDCPublicIP(proxy_url, sessiontoken):
...
 
-def configure_t1(proxy, session_token, t1_id, connection_type):
+def configure_t1(**kwargs):
     """ Configures a Tier1 router as 'ROUTED', 'ISOLATED', or 'NATTED'... Creates a new T1 if it does not exist already."""
-    json_data = {"type": connection_type}
+    if kwargs['objectname'] is None:
+        print("Please use -n to specify the name of the T1 router to be configured.  Consult the help for additional options.")
+        sys.exit()
+    t1_id = kwargs["objectname"]
+    json_data = {"type": kwargs["t1type"]}
```
Comment (kremerpatrick) [inline pyVMC.py:2199]: We should do some argument validation since a user mistake in t1 type will result in an API failure.

**Review by kremerpatrick — CHANGES_REQUESTED** [review]

File: pyVMC.py [hunk 2]
```diff
@@ -2189,18 +2190,146 @@ def getSDDCPublicIP(proxy_url, sessiontoken):
...
+def configure_t1(**kwargs):
     """ Configures a Tier1 router as 'ROUTED', 'ISOLATED', or 'NATTED'... Creates a new T1 if it does not exist already."""
-    json_data = {"type": connection_type}
-    configure_t1_json(proxy, session_token, t1_id, json_data)
-    print(f'Tier1 gateway {t1_id} has been configured as {connection_type}')
+    if kwargs['objectname'] is None or kwargs['t1type'] is None:
+        print("Please use -n to specify the name of the T1 router to be configured, and -t1t or --t1type to specify the type (ROUTED/NATTED/ISOLATED).  Consult the help for additional options.")
+        sys.exit()
+    for i, j in kwargs.items():
+        print(i, j)
```
Comment (kremerpatrick) [inline pyVMC.py:2199]: Is this debugging or intentional?

File: pyVMC.py [hunk 3]
```diff
@@ -2189,18 +2190,146 @@ def getSDDCPublicIP(proxy_url, sessiontoken):
...
+        json_data["advanced_config"]["connectivity"] = f'{kwargs["connectivity"]}'
+    if kwargs['routing_type'] is not None:
+        json_data["type"] = f'{kwargs["routing_type"]}'
+    if kwargs['tier1_id'] is not None:
+        if segment_path == "/infra/tier-1s/cgw":
+            print("This is a fixed segment - you may not alter the connectivity path.  Plese create a 'moveable' segment.")
+        else:
+            json_data["connectivity_path"] = f'/infra/tier-1s/{kwargs["tier1_id"]}'
+    # make the call to the API
+    configure_segment_json(proxy, session_token, segment_path, json_data)
```
Comment (kremerpatrick) [inline pyVMC.py:2309]: I don't see where this one was changed like the others were - no return value?

**Review by kremerpatrick — CHANGES_REQUESTED** [review]

**Comment by fr0gger03** [comment]
AHHH! debugging.


Tom Twyman
***@***.***
804-921-2705

 
 

Sent: Friday, July 15, 2022 at 3:31 PM

**Review by kremerpatrick — APPROVED** [review]

**Pull request merged**